### ISSUE TITLE
[common] Support terminationMessagePath

### ISSUE THREAD

**Issue by zamnuts** [k8s-at-home/library-charts#77]
# Details

**Helm chart name:** `common`

**Describe the solution you'd like:**

Add support for container fields `terminationMessagePath` and `terminationMessagePolicy`. Can be kept flat to follow the `Container` spec, or structured, e.g.:

```
# structured
termination:
  path: '/dev/termination-log' # default
  policy: 'File' # default

# or flat
terminationMessagePath: '/dev/termination-log' # default
terminationMessagePolicy: 'File' # default
```

See docs [Customizing the termination message](https://kubernetes.io/docs/tasks/debug-application-cluster/determine-reason-pod-failure/#customizing-the-termination-message) and [Pod Lifecycle](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#lifecycle-1) (note: container termination is in the "Lifecycle" section, but is not within the `lifecycle` fields).

**Anything else you would like to add:**

When specifying a volume `/dev` as read-only, the container will fail to start as it expects `/dev/termination-log` local to the container to be writable. The only way to mount `/dev` without changing the `terminationMessagePath` is as `readOnly: false`:

```
# works, since the host path is both readable and writable
persistence:
  devfs:
    enabled: true
    type: hostPath
    hostPath: /dev

# does NOT work, termination-log is not writable
persistence:
  devfs:
    enabled: true
    type: hostPath
    hostPath: /dev
    readOnly: true
```

Overriding the termination log path is possible with containers that do not support a termination file or provide a termination log path configuration itself. It would look something like this all together:

```
termination:
  path: '/var/termination-log' # change to `/var`
  policy: 'FallbackToLogsOnError' # additionally, for containers that do not have direct instrumentation, a log policy can be used instead

persistence:
  devfs:
    enabled: true
    type: hostPath
    hostPath: /dev
    readOnly: true
```

**Additional Information:**

Although not related to the container termination configuration, there also exists `terminationGracePeriodSeconds` within the `PodSpec`, see the [Probe section of the PodSpec docs](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/pod-v1/#Probe). Termination grace period is for the entire pod, and not distinct containers within the pod. If using the structured approach, it may cause confusion to include a `gracePeriodSeconds` field to the `termination` structure, but worth considering.

**Comment by bjw-s** [comment]
Thanks for taking the time to raise this issue! Also thanks for all the snippets and links, those are much appreciated!

I am curious what the actual use case for this would be? By that I mean, what is the actual problem (and / or app) you are having that requires this solution?

The reason why I am asking, is that we don't want to be copying the entire Kubernetes spec into the library, when other solutions might be better suited.

**Comment by zamnuts** [comment]
I'm using `smarter-device-manager` to inventory devices in `/dev` and make them available as consumable resources via kubelet's device plugins feature. To perform this task, write access should _not_ be needed by `smarter-device-manager`; write access would be needed by the pods consuming those resources, but that doesn't happen via smarter.

See https://gitlab.com/arm-research/smarter/smarter-device-manager/-/blob/master/main.go#L132 which only performs a directory listing and stat of the "files."

`smarter-device-manager` must run with slightly elevated privilege, and it would be nice to limit the extent of this privilege relative to the host.

Currently `/sys` is limited to read-only, and things work just fine:
```
  sysfs:
    enabled: true
    type: hostPath
    hostPath: /sys
    readOnly: true
```
In fact, `/dev` is also read-only by default at https://github.com/k8s-at-home/charts/blob/233bac5c4797f4689bb141f53449de84a0a946c3/charts/stable/smarter-device-manager/values.yaml#L30-L35 but in k8s 1.20, this doesn't work as expected. I haven't dug into any k8s changes to find the cause.

**Comment by bjw-s** [comment]
Thanks for getting back to us. In the meantime https://github.com/k8s-at-home/charts/issues/1092 was raised as well. We will try to implement a solution to setting these fields.

**Comment by bjw-s** [comment]
This should be resolved once we release `common` version 4.0.0.
